### ISSUE TITLE
`call-non-callable` false positive after `hasattr()` narrowing

### ISSUE THREAD

**Issue by mxr** [astral-sh/ty#3753]
### Summary

Hi, I'm seeing an issue with `hasattr` narrowing. `hasattr` narrowing infers the attribute type as `object` instead of `Unknown`, so attribute calls fail with `call-non-callable`. See repro at https://play.ty.dev/8c9a870a-67b5-4ea9-85a8-ca6d31158f84

```py
import unittest


def example(instance: unittest.TestCase) -> None:
    if hasattr(instance, "make_instance"):
        instance.make_instance(include_optional=False)  # Object of type `object` is not callable (call-non-callable) [Ln 6, Col 9]
```

`mypy` narrowing shows the attribute is `Any`, not `object` (which I believe is the correct behavior) and supports calling the attribute.

**Expected**: no error (attribute type should be `Unknown`)
**Actual**: `call-non-callable` — `object` is not callable

I believe the fix would be something like

```diff
# in types/narrow.rs
-                    // Since `hasattr` only checks if an attribute is readable,
-                    // the type of the protocol member should be a read-only property that returns `object`.
+                    // Since `hasattr` only checks if an attribute is readable and provides no
+                    // information about the attribute's type, use `Unknown` rather than `object`.
+                    // Using `object` prevents callers from calling the attribute.
                     let constraint =
-                        Type::protocol_with_readonly_members(self.db, [(attr, Type::object())]);
+                        Type::protocol_with_readonly_members(self.db, [(attr, Type::unknown())]);
```

### Version

ty 0.0.48 (10b243db0 2026-06-10)

**Comment by sharkdp** [comment]
The `hasattr` narrowing behavior is intentionally strict. There is nothing that indicates that this attribute is callable. Even adding an additional `and callable(instance.make_instance)` check will not help, since that wouldn't give us any information about its signature.

You could either ignore this error (`# ty: ignore` or use a cast to a `Callable` type), or, if possible, restructure your code so that the `make_instance` attribute is typed, e.g. by subclassing `TestCase`. If you have multiple usage sites, you could consider doing something like the following, which wouldn't give you any more type safety, but move the problem to away from the usage site. And you could model the return type of `make_instance`.


```py
from typing import TYPE_CHECKING
import unittest

if TYPE_CHECKING:
    from typing import TypeIs, Protocol

    from ty_extensions import Intersection

    class HasMakeInstance(Protocol):
        def make_instance(self, include_optional: bool) -> WhateverYouWantHere: ...

    def has_make_instance(instance: unittest.TestCase) -> TypeIs[Intersection[unittest.TestCase, HasMakeInstance]]:
        ...
else:
    def has_make_instance(instance):
        return hasattr(instance, "make_instance")


def example(instance: unittest.TestCase) -> None:
    if has_make_instance(instance):
        instance.make_instance(include_optional=False)
```

**Comment by mxr** [comment]
OK thanks that makes sense. Coming from `mypy` I expected the same narrowing behavior however I see the reason why this is more strict. Is there a good place to document these kinds of differences?

**Comment by sharkdp** [comment]
Yes, we have [this page](https://docs.astral.sh/ty/coming-from-mypy-or-pyright/) in our documentation, which could certainly be extended with behavioral changes. That's going to be a lot of work though, and since some of our behaviors are still in flux, I would assume that this is not a high priority at the moment.

**Comment by mxr** [comment]
I see. Would you take a PR on that page even if you aren't ready to formally enumerate all behavioral changes? Something like

```md
## Behavioral differences from mypy

_this list is not exhaustive_

### `hasattr` narrowing

`mypy` narrows `hasattr(obj, 'attr')` to `Any`, allowing the attribute to be used
freely (e.g., called as a function). `ty` narrows to `object` instead, so calling
or indexing the attribute is still an error.

To achieve narrowing, use a `Protocol` with `TypeIs` and `Intersection`: (example below)
```

**Comment by sharkdp** [comment]
Hm, maybe? If we can write that section in a more general way and also talk about other forms of narrowing, e.g. `isinstance(obj, list)` or `if callable(obj)`, which show similar behavioral differences. @AlexWaygood any thoughts on this?

**Comment by AlexWaygood** [comment]
Pyright does something similar: https://microsoft.github.io/pyright/#/mypy-comparison. Users would probably find such a page useful. I'm supportive (we may want to mention differences to pyright too, though, rather than just mypy).
